modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/FrontendDataSet.tsx

```diff
@@ -351,6 +420,75 @@ const FrontendDataSetContent = ({
 		}
 	}
 
+	const handlePopState = () => {
+		const stateFromURL = getStateFromURL(id);
+		const stateUpdates = [];
+		const delta = getDeltaFromURL(stateFromURL);
+
+		if (delta && delta !== paginationDelta) {
+			setPageNumber(1);
+			stateUpdates.push({
+				type: VIEWS_ACTION_TYPES.UPDATE_PAGINATION_DELTA,
+				value: delta,
+			});
+		}
+
+		const filtersFromURL = getFiltersFromURL(stateFromURL);
+
+		if (filtersFromURL) {
+			stateUpdates.push({
+				type: VIEWS_ACTION_TYPES.UPDATE_FILTERS,
+				value: filters?.map((filter: any): any => {
+					const filterFromURL = filtersFromURL.find(
+						(filterFromURL: any) => filterFromURL.id === filter.id
+					);
+
+					if (filterFromURL) {
+						return activateFilter({
+							filter,
+							selectedData: filterFromURL.selectedData,
+						});
+					}
+
+					return deactivateFilter(filter);
+				}),
+			});
+		}
+
+		const viewNameFromURL = getViewNameFromURL(stateFromURL);
+
+		if (viewNameFromURL) {
+			stateUpdates.push({
+				type: VIEWS_ACTION_TYPES.UPDATE_ACTIVE_VIEW,
+				value: viewNameFromURL,
+			});
+		}
+
+		if (stateUpdates.length) {
+			setStateFromURL(stateFromURL);
+			viewsDispatch({
+				type: VIEWS_ACTION_TYPES.BATCH_UPDATE,
+				value: stateUpdates,
+			});
+		}
+	};
+
+	useEffect(() => {
+		const registerEvent =
+			stateInURLSettings === EStateInURLSettings.PUSH &&
+			(!Liferay.SPA || !Liferay.SPA.app);
+
+		if (registerEvent) {
+			window.addEventListener('popstate', handlePopState);
+		}
+
+		return () => {
+			if (registerEvent) {
+				window.removeEventListener('popstate', handlePopState);
+			}
+		};
+	}, [handlePopState]);
```

Comment:
I'm wondering if it'd be more cleaner having a hook to encapsulate all the "state in URL" logic. 

I haven't fully thought this through, but ideally I'd like something like:

```
const [stateInURL, setStateInURL] = useStateInURL(id, stateInURLSettings);
```

So inside here could be this `popstate` useEffect, `stateFromURL` states. 

I'm actually still confused on the difference between `setStateFromURL` and `writeStateInURL` like should those be linked somehow? Like should `setStateFromURL` call `writeStateInURL`? 🤔